posthog/exception_capture.py

```diff
@@ -0,0 +1,61 @@
+import json
+import logging
+import sys
+import threading
+from typing import TYPE_CHECKING
+
+from posthog.exception_utils import exceptions_from_error_tuple, single_exception_from_error_tuple
+
+if TYPE_CHECKING:
+    from posthog.client import Client
+
+
+class ExceptionCapture:
+    # TODO: Add client side rate limiting to prevent spamming the server with exceptions
+
+    log = logging.getLogger("posthog")
+
+    def __init__(self, client: "Client"):
+        self.client = client
+        self.original_excepthook = sys.excepthook
+        sys.excepthook = self.exception_handler
+        threading.excepthook = self.thread_exception_handler
+
+    def exception_handler(self, exc_type, exc_value, exc_traceback):
+        # don't affect default behaviour.
+        self.capture_exception(exc_type, exc_value, exc_traceback)
+        self.original_excepthook(exc_type, exc_value, exc_traceback)
+
+    def thread_exception_handler(self, args):
+        self.capture_exception(args.exc_type, args.exc_value, args.exc_traceback)
+
+    def capture_exception(self, exc_type, exc_value, exc_traceback):
+        try:
+            # if hasattr(sys, "ps1"):
+            #     # Disable the excepthook for interactive Python shells
+            #     return
+
+            # Format stack trace like sentry
+            exception_info = single_exception_from_error_tuple(exc_type, exc_value, exc_traceback)
+            stack_trace = (
+                exception_info["stacktrace"]["frames"]
+                if "stacktrace" in exception_info and exception_info["stacktrace"].get("frames")
+                else None
+            )
+            all_exceptions_with_trace = exceptions_from_error_tuple((exc_type, exc_value, exc_traceback))
+
+            properties = {
+                "$exception_type": all_exceptions_with_trace[0].get("type"),
+                "$exception_message": all_exceptions_with_trace[0].get("value"),
+                "$exception_stack_trace_raw": json.dumps(stack_trace),
```

Comment:
these 3 properties should be inferred from the 1st item of the `all_exceptions_with_trace`, otherwise it's just duplicating data

Reply:
whoops, nice catch, missed this in cleanup! - We can delete `$exception_stack_trace_raw` now imo, its superceded by `$exception_list`

Reply:
(will wait to hear from @daibhin / @pauldambra  before merging incase we want to change names / adjust something here / figure out if for some reason this doesn't work well in js-land)

Reply:
JS also has chained exceptions via the `clause` prop so its technically possible.

Reply:
Thought we might use `$exception_stack_trace_raw` in the Sentry integration but it doesn't look like we do. Happy to remove it in that case